File: services/agent-gateway/vss_agent_gateway/connectors/legacy_chat.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Compatibility connector for the existing VSS OpenAI-shaped chat stream."""

from __future__ import annotations

import hashlib
import json
import threading
import urllib.error
import urllib.request
from collections.abc import Iterator
from threading import Event

from ..config import GatewayConfig
from ..contract import ConnectorEvent, CreateRunRequest
from .base import Connector, ConnectorError


class LegacyChatConnector(Connector):
    def __init__(self, config: GatewayConfig) -> None:
        self._config = config
        self._endpoint = f"{config.backend_url}{config.backend_path}"
        self._lock = threading.RLock()
        self._active_responses: dict[str, object] = {}

    @property
    def protocol(self) -> str:
        return "legacy-chat"

    @staticmethod
    def _session_key(thread_id: str) -> str:
        return f"vss-ui-{hashlib.sha256(thread_id.encode()).hexdigest()[:40]}"

    @staticmethod
    def _step_event(payload: object) -> ConnectorEvent | None:
        if not isinstance(payload, dict):
            return None
        status = str(payload.get("status") or "in_progress").lower()
        data: dict[str, object] = {
            "tool_call_id": str(payload.get("id") or "tool"),
            "name": str(payload.get("name") or "Agent step"),
            "payload": payload.get("payload"),
        }
        if status in {"complete", "completed", "success", "succeeded"}:
            return ConnectorEvent("tool.completed", data)
        if status in {"error", "failed", "failure"}:
            data["error"] = payload.get("error")
            return ConnectorEvent("tool.failed", data)
        return ConnectorEvent("tool.started", data)

    @staticmethod
    def _content(payload: object) -> str | None:
        if not isinstance(payload, dict):
            return None
        choices = payload.get("choices")
        if (
            not isinstance(choices, list)
            or not choices
            or not isinstance(choices[0], dict)
        ):
            return None
        choice = choices[0]
        delta = choice.get("delta")
        message = choice.get("message")
        if isinstance(delta, dict) and isinstance(delta.get("content"), str):
            return delta["content"]
        if isinstance(message, dict) and isinstance(message.get("content"), str):
            return message["content"]
        return None

    def run(
        self,
        request: CreateRunRequest,
        *,
        run_id: str,
        cancel_event: Event,
    ) -> Iterator[ConnectorEvent]:
        messages = request.full_transcript()
        payload = {
            "model": self._config.backend_model,
            "messages": [message.to_chat_message() for message in messages],
            "stream": True,
        }
        headers = {
            "Accept": "text/event-stream",
            "Content-Type": "application/json",
            "Conversation-Id": self._session_key(request.thread_id),
            "User-Agent": "vss-agent-gateway/1.0",
            **self._config.backend_headers,
        }
        if self._config.backend_token:
            headers["Authorization"] = f"Bearer {self._config.backend_token}"
        upstream_request = urllib.request.Request(
            self._endpoint,
            data=json.dumps(
                payload, ensure_ascii=False, separators=(",", ":")
            ).encode(),
            headers=headers,
            method="POST",
        )

        try:
            response = urllib.request.urlopen(
                upstream_request,
                timeout=self._config.request_timeout_seconds,
            )
        except urllib.error.HTTPError as error:
            error.read(64_000)
            raise ConnectorError(
                f"backend returned HTTP {error.code}",
                code="backend_http_error",
                retryable=error.code >= 500,
            ) from error
        except urllib.error.URLError as error:
            raise ConnectorError(
                "backend is unreachable", code="backend_unreachable", retryable=True
            ) from error

        done = False
        with response:
            with self._lock:
                self._active_responses[run_id] = response
            try:
                for raw_line in response:
                    if cancel_event.is_set():
                        return
                    line = raw_line.decode("utf-8", errors="replace").strip()
                    if not line:
                        continue
                    if line.startswith("intermediate_data:"):
                        try:
                            intermediate = json.loads(line.partition(":")[2].strip())
                        except json.JSONDecodeError:
                            continue
                        event = self._step_event(intermediate)
                        if event:
                            yield event
                        continue
                    if not line.startswith("data:"):
                        continue
                    data = line.partition(":")[2].strip()
                    if data == "[DONE]":
                        done = True
                        break
                    try:
                        parsed = json.loads(data)
                    except json.JSONDecodeError:
                        continue
                    content = self._content(parsed)
                    if content:
                        yield ConnectorEvent("message.delta", {"delta": content})
            except (OSError, ValueError) as error:
                if cancel_event.is_set():
                    return
                raise ConnectorError(
                    "backend stream ended unexpectedly",
                    code="backend_stream_error",
                    retryable=True,
                ) from error
            finally:
                with self._lock:
                    self._active_responses.pop(run_id, None)

        if not done and not cancel_event.is_set():
            raise ConnectorError(
                "backend stream ended before [DONE]",
                code="incomplete_backend_stream",
                retryable=True,
            )

    def cancel(self, run_id: str) -> None:
        with self._lock:
            response = self._active_responses.get(run_id)
        close = getattr(response, "close", None)
        if callable(close):
            close()

File: services/ui/apps/nv-metropolis-bp-vss-ui/utils/server/agentAdapter/connectors/legacyChat.ts
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import type { AgentAdapterConfig } from "../config";
import {
  type ConnectorEvent,
  type CreateRunRequest,
  fullTranscript,
} from "../contract";
import { isJsonObject, strictJsonParse } from "../json";
import {
  boundedResponseText,
  linkedTimeoutSignal,
  readLines,
} from "../streams";
import { type Connector, ConnectorError, connectorCapabilities } from "./base";
import { createHash } from "node:crypto";

const asString = (value: unknown, fallback: string): string =>
  typeof value === "string" ? value : fallback;

export class LegacyChatConnector implements Connector {
  readonly protocol = "legacy-chat";
  readonly capabilities = connectorCapabilities(this.protocol);
  private readonly endpoint: string;

  constructor(private readonly config: AgentAdapterConfig) {
    this.endpoint = `${config.backendUrl}${config.backendPath}`;
  }

  private static sessionKey(threadId: string): string {
    return `vss-ui-${createHash("sha256")
      .update(threadId)
      .digest("hex")
      .slice(0, 40)}`;
  }

  private static stepEvent(payload: unknown): ConnectorEvent | null {
    if (!isJsonObject(payload)) return null;
    const status = asString(payload.status, "in_progress").toLowerCase();
    const data = {
      tool_call_id: asString(payload.id, "tool"),
      name: asString(payload.name, "Agent step"),
      payload: payload.payload,
    };
    if (["complete", "completed", "success", "succeeded"].includes(status)) {
      return { type: "tool.completed", data };
    }
    if (["error", "failed", "failure"].includes(status)) {
      return {
        type: "tool.failed",
        data: { ...data, error: payload.error },
      };
    }
    return { type: "tool.started", data };
  }

  private static content(payload: unknown): string | null {
    if (!isJsonObject(payload) || !Array.isArray(payload.choices)) return null;
    const choice = payload.choices[0];
    if (!isJsonObject(choice)) return null;
    if (
      isJsonObject(choice.delta) &&
      typeof choice.delta.content === "string"
    ) {
      return choice.delta.content;
    }
    if (
      isJsonObject(choice.message) &&
      typeof choice.message.content === "string"
    ) {
      return choice.message.content;
    }
    return null;
  }

  async *run(
    request: CreateRunRequest,
    _runId: string,
    signal: AbortSignal
  ): AsyncGenerator<ConnectorEvent> {
    const timeout = linkedTimeoutSignal(signal, this.config.requestTimeoutMs);
    const headers = new Headers({
      Accept: "text/event-stream",
      "Content-Type": "application/json",
      "Conversation-Id": LegacyChatConnector.sessionKey(request.threadId),
      "User-Agent": "vss-next-agent-adapter/1.0",
      ...this.config.backendHeaders,
    });
    if (this.config.backendToken) {
      headers.set("Authorization", `Bearer ${this.config.backendToken}`);
    }
    let response: Response;
    try {
      response = await fetch(this.endpoint, {
        method: "POST",
        headers,
        body: JSON.stringify({
          model: this.config.backendModel,
          messages: fullTranscript(request),
          stream: true,
        }),
        signal: timeout.signal,
      });
    } catch (error) {
      timeout.cleanup();
      if (signal.aborted) return;
      if (timeout.signal.aborted) {
        throw new ConnectorError("backend timed out", "backend_timeout", true, {
          cause: error,
        });
      }
      throw new ConnectorError(
        "backend is unreachable",
        "backend_unreachable",
        true,
        { cause: error }
      );
    }
    if (!response.ok) {
      await boundedResponseText(response, 64_000).catch(() => "");
      timeout.cleanup();
      throw new ConnectorError(
        `backend returned HTTP ${response.status}`,
        "backend_http_error",
        response.status >= 500
      );
    }

    let done = false;
    try {
      for await (const rawLine of readLines(response)) {
        if (signal.aborted) return;
        const line = rawLine.trim();
        if (!line) continue;
        if (line.startsWith("intermediate_data:")) {
          try {
            const event = LegacyChatConnector.stepEvent(
              strictJsonParse(line.slice(line.indexOf(":") + 1).trim())
            );
            if (event) yield event;
          } catch {
            // The legacy stream historically ignored malformed intermediate rows.
          }
          continue;
        }
        if (!line.startsWith("data:")) continue;
        const data = line.slice(line.indexOf(":") + 1).trim();
        if (data === "[DONE]") {
          done = true;
          break;
        }
        try {
          const content = LegacyChatConnector.content(strictJsonParse(data));
          if (content)
            yield { type: "message.delta", data: { delta: content } };
        } catch {
          // Preserve compatibility with the original best-effort legacy parser.
        }
      }
    } catch (error) {
      if (signal.aborted) return;
      throw new ConnectorError(
        timeout.signal.aborted
          ? "backend timed out"
          : "backend stream ended unexpectedly",
        timeout.signal.aborted ? "backend_timeout" : "backend_stream_error",
        true,
        { cause: error }
      );
    } finally {
      timeout.cleanup();
    }
    if (!done && !signal.aborted) {
      throw new ConnectorError(
        "backend stream ended before [DONE]",
        "incomplete_backend_stream",
        true
      );
    }
  }

  cancel(): void {
    // The shared AbortController passed to fetch performs cancellation.
  }
}
